source/grid/grid_tools.cc

```diff
@@ -5883,68 +5888,119 @@ namespace GridTools
-              if (is_valid(box))
-                boxes_and_ranks.emplace_back(box, rank);
+      for (unsigned rank = 0; rank < global_bboxes_to_be_used->size(); ++rank)
+        for (const auto &box : (*global_bboxes_to_be_used)[rank])
+          if (is_valid(box))
+            boxes_and_ranks.emplace_back(box, rank);
 
-          // pack boxes into r-tree
-          const auto tree = pack_rtree(boxes_and_ranks);
+      // pack boxes into r-tree
+      const auto tree = pack_rtree(boxes_and_ranks);
 
-          // loop over all entities
-          for (unsigned int i = 0; i < entities.size(); ++i)
-            {
-              // create a bounding box with tolerance
-              const auto bb =
-                BoundingBox<spacedim>(entities[i]).create_extended(tolerance);
+      // loop over all entities
+      for (unsigned int i = 0; i < entities.size(); ++i)
+        {
+          // create a bounding box with tolerance
+          const auto bb =
+            BoundingBox<spacedim>(entities[i]).create_extended(tolerance);
 
-              // determine ranks potentially owning point/bounding box
-              std::set<unsigned int> my_ranks;
+          // determine ranks potentially owning point/bounding box
+          std::set<unsigned int> my_ranks;
 
-              for (const auto &box_and_rank :
-                   tree | boost::geometry::index::adaptors::queried(
-                            boost::geometry::index::intersects(bb)))
-                my_ranks.insert(box_and_rank.second);
+          for (const auto &box_and_rank :
+               tree | boost::geometry::index::adaptors::queried(
+                        boost::geometry::index::intersects(bb)))
+            my_ranks.insert(box_and_rank.second);
 
-              for (const auto rank : my_ranks)
-                ranks_and_indices.emplace_back(rank, i);
-            }
+          for (const auto rank : my_ranks)
+            ranks_and_indices.emplace_back(rank, i);
         }
-      else
+
+#else
+
+      auto fill_ranks_and_indices =
+        [&ranks_and_indices](
+          const std::vector<int> &                offsets,
+          const std::vector<std::pair<int, int>> &indices_ranks) {
+          for (unsigned int i = 0; i < offsets.size() - 1; ++i)
+            for (int j = offsets[i]; j < offsets[i + 1]; ++j)
+              ranks_and_indices.emplace_back(indices_ranks[j].second, i);
+        };
+
+      unsigned int idx = (global_bboxes.size() == 1) ?
+                           0 :
+                           Utilities::MPI::this_mpi_process(comm);
+      // size of global_bboxes>1
+      if constexpr (spacedim == 2)
+        {
+          // Create query bounding boxes out of entities
+          std::vector<BoundingBox<2>> query_bounding_boxes;
+          for (const auto &entity : entities)
+            query_bounding_boxes.push_back(
+              BoundingBox<2>(entity).create_extended(tolerance));
```

Comment:
Is this the only difference between the 2D and the 3D case?

Reply:
Yes. I unified the two cases in a8e421a170. I didn't do that before that because on my local machine (I have ArborX 1.1) I have linker errors with this variant.